### PR TITLE
SM: 20221101 update

### PR DESCRIPTION
**Pull request by lordlou**
This adds support to most of Varia's 20221101 update. Notably, added Options for:
- Objectives
- Tourian
- RelaxedRoundRobinCF

As well as previously unsupported Options:
- EscapeRando
- RemoveEscapeEnemies
- HideItems

I did some basic tests myself for each options.

Uses updated basepatch and variapatches generated from https://github.com/lordlou/SMBasepatch/tree/update_20221101

### CODE REVIEW

**Comment by Berserker66** [comment]
I'd really like it if all the sys.exit could be replaced with exceptions, because if any of them are encountered by the webhost, then that generation process exits, once all exit the webhost is unable to generate until a restart.

File: worlds/sm/__init__.py [hunk 1]
```diff
@@ -682,6 +699,41 @@ def pre_fill(self):
                 loc.place_locked_item(item)
                 loc.address = loc.item.code = None
 
+    def post_fill(self):
+        self.itemLocs = [
+            ItemLocation(ItemManager.Items[itemLoc.item.type
+                         if isinstance(itemLoc.item, SMItem) and itemLoc.item.type in ItemManager.Items else
+                         'ArchipelagoItem'],
+                         locationsDict[itemLoc.name], itemLoc.item.player, True)
+            for itemLoc in self.multiworld.get_locations() if itemLoc.player == self.player
```
Comment (Berserker66) [inline worlds/sm/__init__.py:708]:
```suggestion
            for itemLoc in self.multiworld.get_locations(self.player)
```

File: worlds/sm/__init__.py [hunk 2]
```diff
@@ -682,6 +699,41 @@ def pre_fill(self):
...
+        ]
+        self.progItemLocs = [
+            ItemLocation(ItemManager.Items[itemLoc.item.type
+                         if isinstance(itemLoc.item, SMItem) and itemLoc.item.type in ItemManager.Items else
+                         'ArchipelagoItem'],
+                         locationsDict[itemLoc.name], itemLoc.item.player, True)
+            for itemLoc in self.multiworld.get_locations() if itemLoc.player == self.player and itemLoc.item.advancement == True
```
Comment (Berserker66) [inline worlds/sm/__init__.py:715]:
```suggestion
            for itemLoc in self.multiworld.get_locations(self.player) itemLoc.item.advancement == True
```

File: worlds/sm/__init__.py [hunk 3]
```diff
@@ -682,6 +699,41 @@ def pre_fill(self):
                 loc.place_locked_item(item)
                 loc.address = loc.item.code = None
 
+    def post_fill(self):
```
Comment (Berserker66) [inline worlds/sm/__init__.py:702]: post_fill runs in parallel to other worlds' post fill and before progression balancing. Just making sure you know.

File: worlds/sm/Options.py [hunk 4]
```diff
@@ -239,6 +244,94 @@ class VariaCustomPreset(OptionList):
...
+    Note: Current percentage of collected items is displayed in the inventory pause menu.
+    Note: Collect 100% items is excluded by default when randomizing the objectives list as it requires you to complete all the objectives.
+    Note: In AP, Items% and areas objectives are counted toward location checks, not items collected or received, except for "collect all upgrades"
+
+    Format as a comma-separated list of objective names: ["kill three G4", "collect 75% items"].
+    A full list of supported objectives can be found at:
+    https://github.com/ArchipelagoMW/Archipelago/blob/main/worlds/sm/utils/objectives.py
+    """
+    display_name = "Objectives"
+    default = "kill all G4"
```
Comment (Alchav) [inline worlds/sm/Options.py:303]: I think this needs to be `default = ["kill all G4"]`? What this puts in the yaml template leads to generation failure

File: worlds/sm/__init__.py (hunk 3 above)
Comment (lordlou) [reply worlds/sm/__init__.py:702]: Yes, it should be good.

**Comment by Berserker66** [comment]
image

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**